[Rachel Tai Ke Jia] iP

rxchell opened the pull request:
# Ollie 

> _Organize with ease, accomplish with Ollie!_ ☺️

### Ollie takes the load off your mind by **managing your tasks for you**!
- Simple, text-based interface
- Quick to learn 
- Convenient to use 
- ~~Slow~~ Efficient

### Features 👍🏼 
- [x] Add a task
- [x] Mark task as completed or not completed
- [x] Find a task by keyword
- [ ] More to come!

### Steps 
1. Download it from [here](https://github.com/rxchell/ip/releases/tag/A-Jar) 👈🏼 
2. Double-click to open
3. Start chatting with `Ollie`! 

```java
Made with Java
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

rxchell committed:
{current situation} Chatbot is named Duke

{why it needs to change} Rename the chatbot to differentiate it

{what is being done about it} Rename the chatbot to Ollie, and add the greeting and exit messages

rxchell committed:
Delete Duke.java

Remove all traces of Duke

rxchell committed:
Merge pull request #1 from rxchell/level-0-rename-greet-exit

Level 0 rename greet exit

rxchell committed:
Make the chatbot echo

**Current Situation**
The chatbot only greets and exits

**Why it needs to change**
The chatbot should echo as well, and it has some personality

**What is being done about it**
* The chatbot returns the commands said by the user
* The chatbot greets and says bye
* The chatbot has cutomised message borders
* Chatbot personality: The chatbot has :) in the messages

rxchell committed:
Merge pull request #2 from rxchell/level-1-echo

Make the chatbot echo and have some personality

rxchell committed:
Add and list tasks

rxchell committed:
make the list of items final
**Current Situation**
* the list of items is not marked as final

**What is done**
* use a fixed size array (e.g., String[100]) for the list of tasks

rxchell committed:
Merge pull request #3 from rxchell/level-2-add-list

Level 2 Add, List

rxchell committed:
Use a class to represent tasks

rxchell committed:
Add Task class

rxchell committed:
Change method of marking task as done

* method to take in a boolean for convenience

rxchell committed:
Merge pull request #4 from rxchell/extenson-a-classes

Add Task Class [Extension-A-Classes]

rxchell committed:
Mark tasks as done or undone

rxchell committed:
Merge pull request #5 from rxchell/level-3-mark-as-done

Mark tasks as done or undone

rxchell committed:
Add ToDos, Events, Deadlines to inherit from Task

**Current Situation**
* Use Inheritance to support multiple task types

  There are multiple types of tasks that have some similarity between them

**Why it needs to change**
* Use the concept of Inheritance to make it more organised

**What is being done about it**
* Implement classes Todo, Deadline and Event classes to inherit from a Task class.

rxchell committed:
Modify param of Deadline constructor

rxchell committed:
Merge pull request #6 from rxchell/extension-A-Inheritance

Add subclasses Todo, Deadline, Event to inherit from Task

rxchell committed:
Track ToDos, Events, Deadlines

rxchell committed:
Merge pull request #7 from rxchell/Level-4-ToDos-Events-Deadlines

Track ToDos, Events, Deadlines

rxchell committed:
Automated Text UI testing

Test using the I/O redirection technique

rxchell committed:
Merge pull request #8 from rxchell/A-TextUiTesting

Automated Text UI testing

rxchell committed:
Exceptions to handle errors in user inputs

rxchell committed:
Merge pull request #10 from rxchell/A-Exceptions

Create exception classes to handle errors in user inputs

rxchell committed:
Handle errors based on user's input

**Current Situation**
* Ollie does not handle all errors

**Why it needs to change**
* Ollie should let the user know what to do when the input lacks certain information

**What is being done about it**
* Each type of task throws a specific error to allow the user to use the correct command

**Why it is done that way**
* Use Exception classes and Error handling for more concise code

**Any other relevant info**
*

rxchell committed:
Add customised messages by Ollie

rxchell committed:
Merge pull request #11 from rxchell/level-5-handle-errors

Handle errors based on user's input

rxchell committed:
Use ArrayList<Task>

rxchell committed:
Merge pull request #13 from rxchell/A-collections

Use ArrayList<Task>

rxchell committed:
Delete task from the list

rxchell committed:
Merge pull request #14 from rxchell/level-6-delete

Delete task from the list

rxchell committed:
Enumeration for TaskType

rxchell committed:
Merge pull request #15 from rxchell/A-Enums

Enumeration for TaskType

rxchell committed:
Test UIText pass

rxchell committed:
Merge pull request #16 from rxchell/A-TextUiTesting

Test UIText pass

rxchell committed:
Save files to /data

rxchell committed:
Merge pull request #17 from rxchell/branch-Level-7

Save the tasks

rxchell committed:
Load stored tasks from ./data/olllie.txt

rxchell committed:
Merge pull request #18 from rxchell/branch-Level-7

Load stored tasks from ./data/olllie.txt

rxchell committed:
Store the dates as MMM dd yyyy format

**Current Situation**
The deadlines and dates are stored as strings.

**Why it needs to change**
The chatbot should understand 2/12/2019 1800 as 2nd of December 2019, 6pm, instead of treating it as just a String.

**What is being done about it**
* Store deadline dates as a java.time.LocalDateTime
* The user inputs the time in the format yyyy-MM-dd HH:mm
* The time is saved and loaded in the format MMM dd yyyy HH:mm

**Why it is done that way**
* It makes it clearer and more use to the user

rxchell committed:
Merge pull request #19 from rxchell/branch-Level-8

Store the dates as MMM dd yyyy format

rxchell committed:
Add Ui class

**Current Situation**
* The interactions with the user is in the Ollie class

**Why it needs to change**
* Refactor the code to extract out closely related code as classes.

**What is being done about it**
* Make a new class Ui for Ollie to interact with the user eg greeting, exit

**Why it is done that way**
* Utilise OOP principle

rxchell committed:
Add Parser Class

**Current Situation**
* The Ollie class is responsible for both parsing user commands and handling interactions

**Why it needs to change**
* Improve the maintainability and readability of the code.
* Parsing commands should be isolated from UI logic, adhering to the single responsibility principle.

**What is being done about it**
* Extract the parsing logic into a new Parser class that handles command interpretation.
* Move the respond method to the Ui class, allowing it to handle all user interactions, including command processing.

**Why it is done that way**
* Improves code organization and future extensibility

rxchell committed:
Merge pull request #20 from rxchell/A-MoreOOP

Additional OOP for Parser and Ui classes

rxchell committed:
Divide classes into packages

**Current Situation**
* Classes are not organised into packages

**Why it needs to change**
* Improve code organization, readability, and maintainability

**What is being done about it**
* Place task-related classes into the ollie.task package.
* Keep the main classes in the ollie package.
* Group custom exception classes under the ollie.exception package.

**Why it is done that way**
* Enhance code modularity
* Ensure a cleaner project architecture

rxchell committed:
Merge pull request #21 from rxchell/A-Packages

Divide classes into packages

rxchell committed:
Merge pull request #22 from rxchell/gradle-support

Add gradle support

rxchell committed:
Add JUnit tests to Event and Exception classes

**Current Situation**
* Classes are not tested using JUnit tests

**Why it needs to change**
* Ensure that the classes work correctly
* Ensure correct exceptions are thrown

**What is being done about it**
* Add JUnit tests for the classes to verify functionality.
* Add tests for custom exception classes to ensure correct behavior and messaging.

**Why it is done that way**
* Ensure code correctness

rxchell committed:
Merge pull request #23 from rxchell/A-JUnit

Add JUnit tests to Event and Exception classes

rxchell committed:
Package the App as a JAR file

**Current Situation**
* No JAR file created

**Why it needs to change**
* It is less easy to distribute without a JAR file

**What is being done about it**
* Package the app as an executable JAR file so that it can be distributed easily.

**Why it is done that way**
* Users can run the command java -jar "ollie.jar"

rxchell committed:
Merge pull request #24 from rxchell/A-Jar

Package the App as a JAR file

rxchell committed:
Add JavaDoc comments to the classes and methods

**Current Situation**
* Not all classes and methods have Javadoc comments

**Why it needs to change**
* The comments help to tell readers what the code does.

**What is being done about it**
* Let's add header comments to at least half of the non-private classes/methods.

**Why it is done that way**
* It makes the code more understandable to others.

rxchell committed:
Tweak the code to comply with the coding standard

Not all the code complies with the coding standard.

Let's change the code based on the Java coding standard here: https://se-education.org/guides/conventions/java/intermediate.html#conditionals

This ensures consistency across the codebase and better maintainability.

rxchell committed:
Add function to find a task by keyword

There is no function to search for a task by a keyword.

Let's
* Add a command "find"
* Users can find the task(s) that contains the keyword

rxchell committed:
Merge pull request #25 from rxchell/branch-A-JavaDoc

Add JavaDoc comments to the classes and methods

rxchell committed:
Merge pull request #26 from rxchell/branch-A-CodingStandard

Tweak the code to comply with the coding standard

rxchell committed:
Merge branch 'master' into branch-Level-9

rxchell committed:
Merge pull request #27 from rxchell/branch-Level-9

Add function to find a task by keyword

rxchell committed:
Add a GUI to the Ollie chatbot

The user interacts with Ollie through the command line. Users should be able to interact with it through an application.

Let's
* use JavaFX in the application
* create  GUI to the Ollie chatbot
* refactor the code to allow the existing functions to work with the GUI

rxchell committed:
Merge pull request #29 from rxchell/branch-Level-10

Add a GUI to the Ollie chatbot

rxchell committed:
Use checkStyle to detect coding style violations.

There is no automatic checks on whether the code violates the coding style.

Let's
* use Checkstyle to check the Java code against a set of style rules.
* change the current code to make sure it adheres to the rules

rxchell committed:
Merge pull request #30 from rxchell/A-CheckStyle

Use checkStyle to detect coding style violations.

rxchell committed:
Edit the code based on Checkstyle

Not all the code abide by the style rules.

Let's
* analyse the checkstyle output by running ./gradlew checkstyleMain checkstyleTest
* edit the code based on the style rules

rxchell committed:
Merge pull request #31 from rxchell/A-CheckStyle

Edit the code based on Checkstyle

rxchell committed:
Update .gitignore to exclude the txt files

The txt files are included when a commit is pushed, but it is unnecessary

Let's 
* Include the .txt files in .gitignore so that they will not be kept track of

rxchell committed:
Merge pull request #32 from rxchell/review

Update .gitignore to exclude the txt files

taggyhan commented on src/main/java/ollie/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+     * @throws OllieException If the description is empty or the deadline is not provided.
+     */
+    @Override
+    public void validateDescription(String command) throws OllieException {
+        String[] parts = command.split(" /by ");
+        if (parts.length != 2) {
+            throw new OllieException("Please enter a name and deadline for the task! ☺");
+        }
+        if (parts[0].trim().isEmpty()) {
+            throw new EmptyDescriptionException("deadline");
+        }
+        if (parts[1].trim().isEmpty()) {
+            throw new OllieException("Please enter a deadline for the task! ☺");
+        }
+    }
```
the name could indicate the purpose of the method is to throw an exception

taggyhan commented on src/main/java/ollie/task/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     * @param taskNumber The task number to unmark as done.
+     * @throws OllieException If the task number is invalid.
+     */
+    public String unmarkTaskAsDone(int taskNumber) throws OllieException {
+        if (taskNumber >= 0 && taskNumber < tasks.size()) {
+            tasks.get(taskNumber).markAsDone(false);
+            storage.saveTasks(tasks);
+            return Ui.unmarkTaskAsDone(tasks.get(taskNumber));
+        } else {
+            return Ui.returnErrorAsString(
+                    new OllieException("Please enter a valid task number within the list ☺"));
+        }
+    }
```
a bit awkward pharsing - seems a little bit ambigious tbh

rxchell replied:
Hello can I check which part is ambiguous?

taggyhan commented on src/main/java/ollie/task/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     * @param keyword The keyword to search for.
+     * @return The list of tasks that contain the keyword.
+     */
+    public String findTasksByKeyword(String keyword) {
+        ArrayList<Task> matchingTasks = tasks.stream()
+                .filter(task -> task.getDescription().toLowerCase().contains(keyword.toLowerCase()))
+                .collect(Collectors.toCollection(ArrayList::new));
+        return Ui.showFindResults(matchingTasks);
+    }
+
+    /**
+     * Deletes a specific task based on the task number.
+     *
+     * @param taskNumber The task number to delete.
+     * @throws OllieException If the task number is invalid.
+     */
+    public String deleteTask(int taskNumber) throws OllieException {
+        if (taskNumber >= 0 && taskNumber < tasks.size()) {
+            Task removedTask = tasks.remove(taskNumber);
+            taskCount = tasks.size();
+            storage.saveTasks(tasks);
+            return Ui.deleteTask(removedTask, taskCount);
+        } else {
+            return Ui.returnErrorAsString(
+                    new OllieException("Please enter a valid task number within the list ☺"));
```
LGTM

taggyhan commented on src/main/java/ollie/task/Todo.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return The Todo task as a string.
+     */
+    @Override
+    public String saveAsString() {
+        return super.saveAsString();
+    }
```
a bit nitpicking but this method seems redundant since it inherits the super class methods

taggyhan reviewed (commented):
Overall naming is generally well-done, i enjoyed reading through the code. Some parts can be improved as stated. LGTM

yyueyii commented on src/main/java/ollie/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+    public static Deadline createTask(String command) throws OllieException {
+        String[] parts = command.substring(8).split(" /by:");
+
+        if (parts.length != 2) {
+            throw new OllieException("""
+                    Please enter in the format:
+                    deadline task_name /by: due_date
+                    Example: deadline assignment /by: 2021-09-30 23:59""");
+        }
```
I like that the you included an example for the LocalDateTime format 👍

yyueyii commented on src/main/java/ollie/ui/DialogBox.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
```
Maybe the method name could be a bit clearer? e.g. swapImageAndText

yyueyii reviewed (commented):
LGTM!The code adheres to the coding standards and is well organised. The JavaDoc comments are detailed and comprehensive. Great job!

rxchell committed:
Edit the code based on code review

There are some areas for improvement in terms of the Javadoc comments and method names.Some methods can be removed as they are not used.

Let's
- Make the Javadoc comments and method names more specific to the method
- Remove the method if the child class has the same method as the parent class

rxchell committed:
Merge pull request #33 from rxchell/branch-A-FullCommitMessage

Edit the code based on code review

rxchell committed:
Add Assertions

Not all assumptions in the code are identified using assertions.

Let's
* Use the assert feature to document important assumptions that should hold at various points in the code.

rxchell committed:
Improve code quality

There is more to be improved in the readability and code practices.

Let's
* Remove redundant methods
* Edit the Javadoc comments to show the what and why
* Refer to the practices for code quality and make other necessary changes
* Test with  ./gradlew checkstyleMain checkstyleTest again

rxchell committed:
Use Streams

Only the find keywords function uses stream. Task parsing does not use streams.

Let's
* Use Java Streams to split and trim task data string components
* Improve the readability and maintainability of the parseTask method

rxchell committed:
Merge pull request #34 from rxchell/branch-A-Assertions

Add Assertions

rxchell committed:
Merge remote-tracking branch 'origin/master' into branch-A-CodeQuality

rxchell committed:
Merge pull request #35 from rxchell/branch-A-CodeQuality

Merge branch 'master' into branch-A-CodeQuality

rxchell committed:
Merge branch 'master' into branch-A-Streams

Let's
* practice merging the master branch on branch-A-Streams
* resolve the merge conflicts

rxchell committed:
Merge pull request #36 from rxchell/branch-A-Streams

Merge branch 'master' into branch-A-Streams

rxchell committed:
Set up Continuous Integration (CI)

There is no CI/CD integration.

Let's
* Add a .yml file into the [root]\.github\workflows folder for GitHub Actions to run the workflow

rxchell committed:
Fix indentation

The switch statement does not abide by the coding style, causing the checkstyle test to fail.
* Let's change the indentation level of the switch and case statements

rxchell committed:
Merge pull request #37 from rxchell/branch-A-CI-1

Set up Continuous Integration (CI)

rxchell committed:
Add friendlier syntax for commands

It may be a hassle for users to type in the full commands.

Let's
* replace the action commands with alias eg "-f" for find

rxchell committed:
Merge pull request #38 from rxchell/branch-C-FriendlierSyntax

Add friendlier syntax for commands

rxchell committed:
Test CI

The previous A-CI tag was not tracked.

Let's
* try the CI again to ensure the Github actions automatically runs the checks

rxchell committed:
Merge pull request #39 from rxchell/test-CI

Test CI

rxchell committed:
Enhance the code using AI tools

The code can be improved.

Let's
* use ChatGPT to improve the code quality
* eg display the error messages properly

rxchell committed:
Fix test cases.

The exception messages were fixed but the test cases were not updated.

Let's
* update the tests to show the correct message

rxchell committed:
Merge pull request #42 from rxchell/branch-A-AiAssisted

Enhance the code using AI tools

rxchell committed:
Add GUI styling.

The GUI can be enhanced.

Let's
* add some CSS styling.
* make the profile pictures circular

rxchell committed:
Enhance GUI styling.

The GUI can look more appealing.

Let's
* add a background image
* change the colours
* make the dialog box rounder

rxchell committed:
Merge pull request #43 from rxchell/branch-A-BetterGui

Make the GUI better

rxchell committed:
Add more error handling

Some cases of incorrect user inputs are not properly handled.

Let's
* properly parse the task number
* identify missing parts of the event or deadline task

rxchell committed:
Merge pull request #44 from rxchell/branch-A-MoreErrorHandling

Add more error handling

rxchell committed:
Add a personality for Ollie

Ollie can be more unique in its responses.

Let's
* make Ollie have a smile emoji
* give examples to guide the user

rxchell committed:
Merge pull request #45 from rxchell/branch-A-Personality

Add a personality for Ollie

rxchell committed:
Add more JUnit tests

Only some methods are tested.

Let's
* add more tests to other methods

rxchell committed:
Add testData.txt

There is no test data for the storage test.

Let's
* add the .txt file for the test data

rxchell committed:
Add a User Guide

Users might not know how to use Ollie.

Let's
* add a user guide

rxchell committed:
Add a User Guide

Users might not know how to use Ollie.

Let's
* add a user guide

rxchell committed:
Merge pull request #46 from rxchell/A-UserGuide

Add a User Guide

rxchell committed:
Create _config.yml

rxchell committed:
Test GFM

The table is not showing in github pages.

Let's
* try adding _config.yml 
* specify GFM markdown

rxchell committed:
Merge pull request #47 from rxchell/pages

Test GFM

rxchell committed:
Update user guide

Some images are not showing in Github pages.

Let's 
* change the source path

rxchell committed:
Update user guide

Some images are not showing in Github pages.

Let's 
* change the source path

rxchell committed:
Update user guide format

The user guide can look better.

Let's 
* ensure all images are shown
* give more specific instructions

rxchell committed:
Updatre user guide 

The installation instructions can be clearer.

Let's
* make the instructions more specific

rxchell committed:
Update User Guide

The user guide can look better.

Let's
* make the first image smaller

rxchell committed:
Update User Guide

There was a mistake in the example.

Let's
* fix the mistake 
* ensure the information is correct

rxchell committed:
Make User Guide clearer

The information can be clearer.

Let's
* include the commands at the start
* add a break between each section

rxchell committed:
Update User Guide

The order should be consistent 

Let's
* make the order of features consistent
